core/src/main/java/org/apache/iceberg/CherryPickFromSnapshot.java

```diff
@@ -0,0 +1,136 @@
+ * snapshot.
+ *
+ * Cherry-picking should apply the exact set of changes that were done in the original commit.
+ *  - All added files should be added to the new version.
+ *  - Todo: If files were deleted, then those files must still exist in the data set.
+ *  - Does not support Overwrite operations currently. Overwrites are considered as conflicts.
+ *
+ */
+class CherryPickFromSnapshot extends MergingSnapshotProducer<CherryPick> implements CherryPick {
+  private final TableOperations ops;
+  private TableMetadata base;
+  private Long cherryPickSnapshotId = null;
+
+  CherryPickFromSnapshot(TableOperations ops) {
+    super(ops);
+    this.ops = ops;
+    this.base = ops.refresh();
+  }
+
+  @Override
+  protected CherryPick self() {
+    return this;
+  }
+
+  /**
+   * We only cherry pick for appends right now
+   * @return
+   */
+  @Override
+  protected String operation() {
+    Snapshot cherryPickSnapshot = base.snapshot(cherryPickSnapshotId);
+    return cherryPickSnapshot.operation();
+  }
+
+  @Override
+  public CherryPickFromSnapshot cherrypick(long snapshotId) {
+    Preconditions.checkArgument(base.snapshot(snapshotId) != null,
+        "Cannot cherry pick unknown snapshot id: %s", snapshotId);
+
+    this.cherryPickSnapshotId = snapshotId;
+    return this;
+  }
+
+  /**
+   * Apply the pending changes and return the uncommitted changes for validation.
+   * <p>
+   * This does not result in a permanent update.
+   *
+   * @return the uncommitted changes that would be committed by calling {@link #commit()}
+   * @throws ValidationException      If the pending changes cannot be applied to the current metadata
+   * @throws IllegalArgumentException If the pending changes are conflicting or invalid
+   */
+  @Override
+  public Snapshot apply() {
+    ValidationException.check(cherryPickSnapshotId != null,
+        "Cannot cherry pick unknown version: call cherrypick");
+
+    Snapshot cherryPickSnapshot = base.snapshot(cherryPickSnapshotId);
+    String wapId = stagedWapId(cherryPickSnapshot);
+    ValidationException.check(!base.isWapIdPublished(Long.parseLong(wapId)),
+        "Duplicate request to cherry pick wap id that was published already: %s", wapId);
+
+    // only append operations are currently supported
+    if (!cherryPickSnapshot.operation().equals(DataOperations.APPEND)) {
+      throw new UnsupportedOperationException("Can cherry pick only append operations");
+    }
+
+    // Todo:
+    //  - Check if files to be deleted exist in current snapshot,
+    //    ignore those files or reject incoming snapshot entirely?
+    //  - Check if there are overwrites, ignore those files or reject incoming snapshot entirely?
+
+    for (DataFile addedFile : cherryPickSnapshot.addedFiles()) {
+      add(addedFile);
+    }
+    set(SnapshotSummary.PUBLISHED_WAP_ID_PROP, wapId);
+    Snapshot outputSnapshot = super.apply();
+    TableMetadata updated = base.addStagedSnapshot(outputSnapshot);
+    ops.commit(base, updated);
```

Comment:
The commit should not be done in the apply method.

Reply:
done.